### PR TITLE
nydus-image: Fix blob unpacking bug in specified blob directory

### LINKED ISSUE #1340: Some problems about Nydus unpack func supporting blob dir
Some day ago ,I see the branch [imeoer:fix-unpack-blob-dir](https://github.com/imeoer/image-service/tree/fix-unpack-blob-dir)

about nydus unpack func supporting blob dir, then i try to use it ,and i find some problem below:

**1. The branch can build successfully, but when i run any cmd about nydus, it will print a same err(see the pic):**

image

then i annotate the code in 698~699 & 5 lines(see the pic)
image
image

then i can run nydus create and unpack normally.

**2. nydus unpack data seem like inconsistent**

I run a set of cmd from nydus create raw file dir to nydus unpack(see the below cmds):

>./nydus-image create -v 5 -t dir-rafs -D /Users/terrill/Documents/urchin/image-service-blob-dir/image-service-fix-unpack-blob-dir/tests/echarts-image /Users/terrill/Documents/urchin/image-service-blob-dir/image-service-fix-unpack-blob-dir/tests/echarts
[2023-06-27 18:39:43.730654 +08:00] INFO successfully built RAFS filesystem: 
meta blob path: /Users/terrill/Documents/urchin/image-service-blob-dir/image-service-fix-unpack-blob-dir/tests/echarts-image/d78b4b1bfb82844c408ae5f64659f78d221c5df3abe3a4ba670e3b9822d1655b
data blob size: 0xc0e1
data blobs: ["fa964c10cd36503b3bbd0d8218ec03d132214a51b0e4b03d2684770290306bce"]

>./nydus-image create -v 5 -t dir-rafs --chunk-dict /Users/terrill/Documents/urchin/image-service-blob-dir/image-service-fix-unpack-blob-dir/tests/echarts-image/meta -D /Users/terrill/Documents/urchin/image-service-blob-dir/image-service-fix-unpack-blob-dir/tests/echarts-image  /Users/terrill/Documents/urchin/zhangshuiyong/urfs/tests/echarts-add
[2023-06-27 18:42:13.087824 +08:00] INFO RAFS v5 super block features: HASH_BLAKE3 | EXPLICIT_UID_GID | COMPRESSION_ZSTD
[2023-06-27 18:42:13.111437 +08:00] INFO successfully built RAFS filesystem: 
meta blob path: /Users/terrill/Documents/urchin/image-service-blob-dir/image-service-fix-unpack-blob-dir/tests/echarts-image/2eec9becfcfebfb3a58918c3920b990a05a2af95d89037be29501453e0433b15
data blob size: 0xc0e1
data blobs: ["5ee38535110c9470daeb434efbace48b15dfb7c08878cc9effa2bc0b29638a0d", "fa964c10cd36503b3bbd0d8218ec03d132214a51b0e4b03d2684770290306bce"]

>./nydus-image unpack --blob-dir /Users/terrill/Documents/urchin/image-service-blob-dir/image-service-fix-unpack-blob-dir/tests/echarts-image --output ./echarts-add.tar  /Users/terrill/Documents/urchin/image-service-blob-dir/image-service-fix-unpack-blob-dir/tests/echarts-image/meta2
[2023-06-27 18:44:09.327079 +08:00] INFO RAFS v5 super block features: HASH_BLAKE3 | EXPLICIT_UID_GID | COMPRESSION_ZSTD

**then I untar the echarts-add.tar and get the unpack file md5 hash , i found the hash is diff from raw file**

> raw file md5 hash

image

> file md5 hash after unpack 
image

**for reappear this problem conveniently, I upload the source file dir of nydus create(1 image file in dir named 'echarts') & nydus create --chunk-dict(2 image file in dir named 'echarts-add')**

>1 image file in dir named 'echarts'

echarts

> 2 image file in dir named 'echarts-add'

avatar
echarts

### PR DESCRIPTION
**Pull request by An-DJ**
## Relevant Issue (if applicable)
Close #1340 

## Details
Fix the `unpack` bug in multi blobs situation. (Based on the unmerged work in #1338)

## Types of changes

_What types of changes does your PullRequest introduce? Put an `x` in all the boxes that apply:_

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Go over all the following points, and put an `x` in all the boxes that apply._

- [X] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.

### CODE REVIEW

**Comment by An-DJ** [comment]
Follow the steps in #1340 to reproduce that issue and verify if it is fixed:

1. Prepare the data:
``` shell
# create a directory to store only one file `echarts.png`
mkdir echarts

# create a directory to store two file `avatar.png` and `echarts.png`
mkdir echarts-add

# create a directory to store created images (meta, blob)
mkdir echarts-image

# download images
wget https://user-images.githubusercontent.com/1655233/249319028-009b32a0-1563-465c-867e-7623d1b9275d.png -O echarts/echarts.png
wget https://user-images.githubusercontent.com/1655233/249319027-5dcdaf14-631c-46c4-ac11-58b1d4c04a0f.png -O echarts-add/avatar.png
wget https://user-images.githubusercontent.com/1655233/249319028-009b32a0-1563-465c-867e-7623d1b9275d.png -O echarts-add/echarts.png
```

2. Create the first one
``` shell
./nydus-image create -v 5 -t dir-rafs -D echarts-image echarts
[2023-09-17 15:10:14.890232 +00:00] INFO successfully built RAFS filesystem:
meta blob path: echarts-image/59016ae5c245fd768f85cf6eeab74da0d88cd00d9f2df5d3627e6f0177ba38d9
data blob size: 0xc0e1
data blobs: ["fa964c10cd36503b3bbd0d8218ec03d132214a51b0e4b03d2684770290306bce"]
```

3. Create the second one
```shell
./nydus-image create -v 5 -t dir-rafs --chunk-dict echarts-image/59016ae5c245fd768f85cf6eeab74da0d88cd00d9f2df5d3627e6f0177ba38d9 -D echarts-image echarts-add
[2023-09-17 15:12:04.145951 +00:00] INFO RAFS v5 super block features: HASH_BLAKE3 | EXPLICIT_UID_GID | COMPRESSION_ZSTD
[2023-09-17 15:12:04.153042 +00:00] INFO successfully built RAFS filesystem:
meta blob path: echarts-image/86235ef91901df9392244c0f973c118530519f5b932bed33671aa9775e45a5e9
data blob size: 0xc0e1
data blobs: ["5ee38535110c9470daeb434efbace48b15dfb7c08878cc9effa2bc0b29638a0d", "fa964c10cd36503b3bbd0d8218ec03d132214a51b0e4b03d2684770290306bce"]
```

4. Do unpack
```shell
./nydus-image unpack --blob-dir echarts-image --output ./unpack-files.tar echarts-image/86235ef91901df9392244c0f973
c118530519f5b932bed33671aa9775e45a5e9
[2023-09-17 15:13:23.384728 +00:00] INFO RAFS v5 super block features: HASH_BLAKE3 | EXPLICIT_UID_GID | COMPRESSION_ZSTD
```

5. Untar file and compare the checksum:

```shell
$ tar xvf unpack-files.tar
avatar.png
echarts.png

$ md5sum avatar.png
8d852403b0badc7731e8b5aa7bd3ad36  avatar.png
$ md5sum echarts-add/avatar.png
8d852403b0badc7731e8b5aa7bd3ad36  echarts-add/avatar.png

$ md5sum echarts.png
c747eb230ad3439052e001187b6a216d  echarts.png
$ md5sum echarts-add/echarts.png
c747eb230ad3439052e001187b6a216d  echarts-add/echarts.png
```

File: src/bin/nydus-image/main.rs [hunk 1]
```diff
@@ -1321,7 +1321,10 @@ impl Command {
 
                 Some(Arc::new(local_fs))
             }
-            None => Self::get_backend(matches, "unpacker").ok(),
+            None => Some(BlobFactory::new_backend(
+                config.backend.as_ref().unwrap(),
```
Comment (imeoer) [inline src/bin/nydus-image/main.rs:1325]: It's not safe to unpack, let's make sure that these options must be specified as one of them: `"backend-config"`, `"blob-dir"`, `"blob"`.

**Comment by imeoer** [comment]
Great job, thanks for the PR!

File: src/bin/nydus-image/main.rs (hunk 1 above)
Comment (An-DJ) [reply src/bin/nydus-image/main.rs]: Done.

**Review by adam3q — COMMENTED** [review]
Please organize your commits before merging. Our repo uses rebase merging, which will submit your original commit information to the master branch.

**Comment by An-DJ** [comment]
> Please organize your commits before merging. Our repo uses rebase merging, which will submit your original commit information to the master branch.

@adamqqqplay Done.

File: src/bin/nydus-image/main.rs [hunk 2]
```diff
@@ -1321,7 +1321,19 @@ impl Command {
 
                 Some(Arc::new(local_fs))
             }
-            None => Self::get_backend(matches, "unpacker").ok(),
+            None => {
+                if config.backend.is_some() {
```
Comment (imeoer) [inline src/bin/nydus-image/main.rs:1325]: How about:

```
if let Some(backend) = config.backend {
  Some(BlobFactory::new_backend(
      backend.as_ref(),
      "unpacker",
  )?)
}
```

**Review by imeoer — COMMENTED** [review]
others LGTM! Thanks!

File: src/bin/nydus-image/main.rs (hunk 2 above)
Comment (An-DJ) [reply src/bin/nydus-image/main.rs]: Done in recent commit.

**Review by imeoer — APPROVED** [review]

**Pull request merged**